synapse/storage/databases/main/event_push_actions.py

```diff
@@ -134,48 +153,100 @@ def _get_unread_counts_by_receipt_txn(
-            " user_id = ?"
-            " AND room_id = ?"
-            " AND stream_ordering > ?"
+        # We need to look specifically for events with notif = 1 because otherwise we'll
+        # count unread events (from MSC2654) that might not notify.
+        notify_count = self._get_count_from_push_actions_txn(
+            txn=txn,
+            user_id=user_id,
+            room_id=room_id,
+            stream_ordering=stream_ordering,
+            push_actions_column="notif",
+            push_summary_column="notif_count",
         )
 
-        txn.execute(sql, (user_id, room_id, stream_ordering))
-        row = txn.fetchone()
-        notify_count = row[0] if row else 0
+        # Now get the number of highlights
+        highlight_count = self._get_count_from_push_actions_txn(
+            txn=txn,
+            user_id=user_id,
+            room_id=room_id,
+            stream_ordering=stream_ordering,
+            push_actions_column="highlight",
+            push_summary_column=None,
+        )
+
+        # Finally, get the number of unread messages
+        unread_count = self._get_count_from_push_actions_txn(
+            txn=txn,
+            user_id=user_id,
+            room_id=room_id,
+            stream_ordering=stream_ordering,
+            push_actions_column="unread",
+            push_summary_column="unread_count",
+        )
+
+        return {
+            "notify_count": notify_count,
+            "unread_count": unread_count,
+            "highlight_count": highlight_count,
+        }
+
+    def _get_count_from_push_actions_txn(
+        self,
+        txn: LoggingTransaction,
+        user_id: str,
+        room_id: str,
+        stream_ordering: int,
+        push_actions_column: str,
+        push_summary_column: Optional[str],
+    ) -> int:
+        """Counts the number of rows in event_push_actions with the given flag set to
+        true, and adds it up with its matching count in event_push_summary if any.
 
-        txn.execute(
-            """
-            SELECT notif_count FROM event_push_summary
-            WHERE room_id = ? AND user_id = ? AND stream_ordering > ?
-        """,
-            (room_id, user_id, stream_ordering),
-        )
-        rows = txn.fetchall()
-        if rows:
-            notify_count += rows[0][0]
+        Args:
+            user_id: The user to calculate the count for.
+            room_id: The room to calculate the count for.
+            stream_ordering: The stream ordering to use in the conditional clause when
+                querying from event_push_actions.
+            push_actions_column: The column to filter by when querying from
+                event_push_actions. The filtering will be done on the condition
+                "[column] = 1".
+            push_summary_column: The count in event_push_summary to add the results from
+                the first query to. None if there is no count in the event_push_summary
+                table to add the results to.
```

Comment:
> None if there is no count in the event_push_summary table to add the results to.

this implies that we're adding results to the event_push_summary table... which doesn't seem right.